File: leetcode/medium/RottingOranges.py
from typing import List
from collections import deque


class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        """
        Calculates the minimum time (minutes) required to rot all fresh oranges
        in a grid.

        This function uses a Multi-Source Breadth-First Search (BFS) to simulate
        the rotting process layer by layer, where each layer represents one minute.

        Grid values are defined as:
        - 0: Empty cell
        - 1: Fresh orange
        - 2: Rotten orange

        Args:
            grid: An m x n grid representing the state of oranges.

        Returns:
            - The minimum minutes required to rot all oranges.
            - -1 if some fresh oranges are unreachable and cannot be rotted.
            -  0 if there are no fresh oranges to begin with.
        """

        # --- 1. Initialization ---

        # Check for an empty grid or a grid with no columns
        if not grid or not grid[0]:
            return 0

        rows = len(grid)
        cols = len(grid[0])

        # Define constants for grid values (improves readability)
        FRESH = 1
        ROTTEN = 2

        # 4-directional movements: up, down, left, right
        DIRECTIONS = [(-1, 0), (1, 0), (0, -1), (0, 1)]

        queue = deque()
        fresh_oranges_count = 0

        # --- 2. First Pass: Find Sources and Count Fresh Oranges ---
        
        # Scan the grid ONCE to:
        # 1. Add all initially rotten oranges (sources) to the queue.
        # 2. Count the total number of fresh oranges.
        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == ROTTEN:
                    # Add rotten oranges as starting points for the BFS
                    queue.append((r, c))
                elif grid[r][c] == FRESH:
                    # Keep track of the total fresh oranges
                    fresh_oranges_count += 1

        # --- 3. Handle Edge Case: No Fresh Oranges ---
        
        # If there are no fresh oranges, no time is needed.
        if fresh_oranges_count == 0:
            return 0

        minutes_elapsed = 0

        # --- 4. Multi-Source BFS Traversal (Simulate Rotting) ---

        # Continue the BFS as long as there are:
        # 1. Rotten oranges still in the queue to spread rot.
        # 2. Fresh oranges left to be rotted.
        while queue and fresh_oranges_count > 0:

            # 'size' represents all oranges rotting at this specific minute.
            # Processing level by level ensures we track time correctly.
            size = len(queue)

            for _ in range(size):
                row, col = queue.popleft()

                # Check all 4 neighbors
                for dr, dc in DIRECTIONS:
                    nr, nc = row + dr, col + dc  # (new_row, new_col)

                    # Check if the neighbor is in bounds AND is a fresh orange
                    if 0 <= nr < rows and 0 <= nc < cols and grid[nr][nc] == FRESH:
                        
                        # 1. Rot the orange
                        grid[nr][nc] = ROTTEN
                        
                        # 2. Decrement the fresh count
                        fresh_oranges_count -= 1
                        
                        # 3. Add it to the queue to spread rot in the *next* minute
                        queue.append((nr, nc))

            # After processing all nodes at the current level (minute),
            # increment the minute counter.
            minutes_elapsed += 1

        # --- 5. Final Result ---

        # After the BFS, if fresh_oranges_count is 0, all oranges were rotted.
        # Otherwise, some were unreachable.
        return minutes_elapsed if fresh_oranges_count == 0 else -1

File: leetcode/medium/RottingOranges.java
import java.util.ArrayDeque;
import java.util.Queue;

class Solution {
    /**
     * Finds the minimum number of minutes until all fresh oranges are rotten.
     * The rot spreads to adjacent fresh oranges every minute.
     * * @param grid The grid representing the state: 0 (empty), 1 (fresh), 2 (rotten).
     * @return The minimum number of minutes, or -1 if some fresh oranges can never rot.
     */
    public int orangesRotting(int[][] grid) {
        if (grid == null || grid.length == 0) {
            return 0;
        }

        int rows = grid.length;
        int cols = grid[0].length;
        // Queue stores the coordinates of rotten oranges: [row, col]
        Queue<int[]> queue = new ArrayDeque<>();
        int freshOranges = 0;

        // 1. Initialize the queue with all initially rotten oranges (2) 
        //    and count the total number of fresh oranges (1).
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                if (grid[r][c] == 2) {
                    queue.offer(new int[]{r, c});
                } else if (grid[r][c] == 1) {
                    freshOranges++;
                }
            }
        }

        // If there are no fresh oranges to begin with, the time is 0.
        if (freshOranges == 0) {
            return 0;
        }

        // Stores the elapsed time (number of minutes).
        int minutes = 0;

        // The four possible directions: up, down, left, right.
        int[][] directions = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};

        // 2. Perform BFS. The queue size at the start of the loop represents 
        //    all oranges that will rot a *new* set of fresh oranges in the current minute.
        while (!queue.isEmpty()) {
            int size = queue.size();
            boolean newlyRotten = false; // Flag to track if any fresh orange rotted in this minute

            for (int i = 0; i < size; i++) {
                int[] rotten = queue.poll();
                int r = rotten[0];
                int c = rotten[1];

                // Check the four neighbors
                for (int[] dir : directions) {
                    int nr = r + dir[0];
                    int nc = c + dir[1];

                    // Check boundaries and if the neighbor is a fresh orange (1)
                    if (nr >= 0 && nr < rows && nc >= 0 && nc < cols && grid[nr][nc] == 1) {
                        // The orange rots
                        grid[nr][nc] = 2;
                        queue.offer(new int[]{nr, nc});
                        freshOranges--;
                        newlyRotten = true;
                    }
                }
            }
            
            // Increment the minute count only if rot successfully spread in this level.
            // This prevents counting an extra minute at the very end when the queue is populated
            // but all fresh oranges have already rotted.
            if (newlyRotten) {
                minutes++;
            }
        }

        // 3. Final Check
        // If freshOranges is 0, all oranges rotted, return the total time.
        // Otherwise, some fresh oranges are unreachable, return -1.
        return freshOranges == 0 ? minutes : -1;
    }
}